kong/plugins/http-log/handler.lua

```diff
@@ -71,67 +108,50 @@ local function send_payload(self, conf, payload)
   local parsed_url = parse_url(http_endpoint)
   local host = parsed_url.host
   local port = tonumber(parsed_url.port)
+  local userinfo = parsed_url.userinfo
 
   local httpc = http.new()
   httpc:set_timeout(timeout)
 
-  table_clear(headers_cache)
+  local headers = {
+    ["Host"] = host,
+    ["Content-Type"] = content_type,
+    ["Content-Length"] = content_length,
+    ["Authorization"] = userinfo and "Basic " .. encode_base64(userinfo) or nil
+  }
   if conf.headers then
     for h, v in pairs(conf.headers) do
-      headers_cache[h] = v
+      headers[h] = headers[h] or v -- don't override Host, Content-Type, Content-Length, Authorization
     end
   end
 
-  headers_cache["Host"] = parsed_url.host
-  headers_cache["Content-Type"] = content_type
-  headers_cache["Content-Length"] = #payload
-  if parsed_url.userinfo then
-    headers_cache["Authorization"] = "Basic " .. encode_base64(parsed_url.userinfo)
-  end
+  local upstream_url = fmt("%s://%s:%d%s", parsed_url.scheme, host, port, parsed_url.path)
 
-  params_cache.method = method
-  params_cache.body = payload
-  params_cache.keepalive_timeout = keepalive
-
-  local url = fmt("%s://%s:%d%s", parsed_url.scheme, parsed_url.host, parsed_url.port, parsed_url.path)
-
-  -- note: `httpc:request` makes a deep copy of `params_cache`, so it will be
-  -- fine to reuse the table here
-  local res, err = httpc:request_uri(url, params_cache)
+  local res, err = httpc:request_uri(upstream_url, {
+    method = method,
+    headers = headers,
+    body = payload,
+    keepalive_timeout = keepalive,
+    ssl_verify = false,
+  })
   if not res then
     return nil, "failed request to " .. host .. ":" .. tostring(port) .. ": " .. err
   end
 
   -- always read response body, even if we discard it without using it on success
   local response_body = res.body
-  local success = res.status < 400
-  local err_msg
-
-  if not success then
-    err_msg = "request to " .. host .. ":" .. tostring(port) ..
-              " returned status code " .. tostring(res.status) .. " and body " ..
-              response_body
-  end
 
-  return success, err_msg
-end
-
-
-local function json_array_concat(entries)
-  return "[" .. concat(entries, ",") .. "]"
-end
+  kong.log.debug(fmt("http-log sent data to upstream, %s:%s HTTP status %d",
+    host, port, res.status))
 
+  if res.status < 400 then
+    return true
```

Comment:
3xx considered acceptable?

Reply:
I've changed this so that 3xx is considered to be a failure and added a CHANGELOG entry to describe the change.  A 3xx HTTP response code likely indicates a configuration error and should not be interpreted as a success indicator.  Is piggy-backing this breaking change onto the queue changes the right thing to do or should this rather be a separate bug?

Reply:
I'm going to consider this change to be a bug fix that will be piggy-backed onto the queue changes.